### ISSUE TITLE
[V2] pydantic.Field function does not accept Decimal as multiple_of parameter

### ISSUE THREAD

**Issue by AlexanderPodorov** [pydantic/pydantic#6885]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

`pydantic.Field` function does not accept `Decimal` as `multiple_of` parameter. However passing `Decimal` value works fine at runtime.

### Example Code

```Python
from pydantic import Field

f = Field(max_digits=8, decimal_places=2, gt=0, multiple_of=Decimal('0.01'))
# pyright shows: Argument of type "Decimal" cannot be assigned to parameter "multiple_of" of type "float | None" in function "Field"
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
        pydantic-core version: 2.4.0
          pydantic-core build: profile=release pgo=false mimalloc=true
               python version: 3.11.4 (v3.11.4:d2340ef257, Jun  6 2023, 19:15:51) [Clang 13.0.0 (clang-1300.0.29.30)]
                     platform: macOS-13.0.1-arm64-arm-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
(.venv) alexander@Alexander-MacBook-Pro versatrans-api %
```

Selected Assignee: @lig

**Comment by dmontagu** [comment]
PR welcome to update the type annotation

**Comment by AlexanderPodorov** [comment]
Yeah, not sure, but just adding `Decimal` to the type annotation works well for me. Not sure is it acceptable/comprehensive solution though.

**Comment by dmontagu** [comment]
Looking at the relevant code, I see we have `gt`, `lt`, etc. also annotated as `float | None`, but @adriangb I believe we now support anything with a `__gt__` or whatever. Maybe we should change these annotations to be some sort of protocol thing? Happy for `multiple_of` to be `float | Decimal | None` if there's no good protocol for that, but for the others it seems like it might be reasonable?

**Comment by AlexanderPodorov** [comment]
For completeness sake I would suggest to add `int` as well, so that it turns into `int | float | Decimal | None`.
If we are looking for the protocol, maybe `numbers.Number` would work, however maybe too broad, since it will most likely accept `Fraction` or `Complex` or other exotic types.

**Comment by AlexanderPodorov** [comment]
Here is my use case:
```python
from typing import Annotated
from decimal import Decimal
from pydantic import Field, BaseModel

PositiveDecimal = Annotated[
    Decimal,
    Field(max_digits=8, decimal_places=2, gt=0, multiple_of=0.01)  # Decimal('0.01')
]

class Test(BaseModel):
    f: PositiveDecimal

t = Test(f=Decimal('1.00'))
# Throws: Input should be a multiple of 0.01 [type=multiple_of, input_value=Decimal('1.00'), input_type=Decimal]
# Works just fine with Decimal('0.01')
```
Technically the type inside `Annotated` and `Field` arguments `gt`, `multiple_of`, etc. should be of the same type.

**Comment by dmontagu** [comment]
I think actually in principle we can allow arbitrary comparable stuff in `gt` etc., not just numbers. Which was my point about protocols for those fields, where I don't think that makes sense for `multiple_of`

**Comment by AlexanderPodorov** [comment]
Yeah, for instance strings support `gt`, which does not make sense in our context.

**Comment by Viicos** [comment]
> For completeness sake I would suggest to add `int` as well, so that it turns into `int | float | Decimal | None`.

IIRC `float` matches `int` and `float` when dealing with types. But it can be set explicitly as well for clarity

**Comment by Viicos** [comment]
Looking at how `multiple_of` is handled for "arbitrary types" (that can still be validated by Pydantic, but different from the usual `int`, `float`, etc):

https://github.com/pydantic/pydantic/blob/1e413857f48a807d3c81281b52c571f07c7077ee/pydantic/_internal/_validators.py#L281-L283

This would still play well with `multiple_of` being a `Decimal` (`3 % Decimal('3')` gives `Decimal('0')`, and `Decimal('0') == 0` gives `True`). However, I'll have to check how it is handled when validation is done in pydantic-core.

**Comment by sydney-runkle** [comment]
This seems to now be working on `main`! Closing as resolved, accordingly!

**Comment by Viicos** [comment]
The issue isn't solved, we still don't document `Decimal` as a valid type and we should double check how it behaves in `pydantic-core` as per [my comment](https://github.com/pydantic/pydantic/issues/6885#issuecomment-2264949386).

**Comment by GinToFim** [comment]
I'd like to work on this. Could you assign it to me?
planning to update the `multiple_of` type hint to accept `Decimal`
